contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -2880,16 +2880,181 @@ add_entry_to_bbf_schema(const char *schema_name,
+
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+									AccessShareLock);
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(schema_name));
+	ScanKeyEntryInitialize(&scanKey[2], 0,
+				Anum_bbf_schema_perms_object_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(object_name));
+	ScanKeyEntryInitialize(&scanKey[3], 0,
+				Anum_bbf_schema_perms_grantee,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(grantee));
+
+	scan = systable_beginscan(bbf_schema_rel,
+				get_bbf_schema_perms_idx_oid(),
+				true, NULL, 4, scanKey);
+	tuple_bbf_schema = systable_getnext(scan);
+
+	if (HeapTupleIsValid(tuple_bbf_schema))
+	{
+		Form_bbf_schema_perms bbf_schema = (Form_bbf_schema_perms) GETSTRUCT(tuple_bbf_schema);
+		permission = bbf_schema->permission;
+	}
+
+	systable_endscan(scan);
+	table_close(bbf_schema_rel, AccessShareLock);
+	return permission;
+}
+
+/*
+ * Update permission column in the catalog when privilege on an object is changed.
+ */
+void
+update_bbf_schema_privilege(const char *schema_name,
+						const char *object_name,
+						int16 new_priv,
+						int16 old_priv,
+						const char *grantee,
+						const char *object_type,
+						bool is_grant)
+{
+	Relation	bbf_schema_rel;
+	HeapTuple	tuple_bbf_schema;
+	TupleDesc	bbf_schema_dsc;
+	HeapTuple	new_tuple;
+	ScanKeyData scanKey[5];
+	TableScanDesc scan;
+	int16	dbid = get_cur_db_id();
+	int16 permission = 0;
+	Datum		new_record_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS];
+	bool		new_record_nulls_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS];
+	bool		new_record_repl_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS];
+
+	if(is_grant)
+	{
+		/*
+		 * In case of GRANT, we add the new privilege along with the previous privilege in the column.
+		 */
+		permission = new_priv | old_priv;
+	}
+	else
+	{
+		/*
+		 * In case of REVOKE, we remove the new privilege and keep the previous privilege as it is.
+		 */
+		permission = ~new_priv & old_priv;
+	}
+
+	if(permission == 0)
+	{
+		del_from_bbf_schema(schema_name, object_name, grantee);
+		return;
+	}
+
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+							RowExclusiveLock);
+
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(schema_name));
+	ScanKeyEntryInitialize(&scanKey[2], 0,
```

Comment:
We need to scan based on object type too!!!

How are you going to distinguish between a table and procedure?
```
1> create table t (a int)
2> go

1> create procedure t as select 1
2> go
```